### ISSUE TITLE
Add AS3 function to scale whole Stage content

### ISSUE THREAD

**Issue by LilithRainbows** [airsdk/Adobe-Runtime-Support#2425]
It would be extremely useful to add a function in AS3 to scale whole Stage content to a custom value (that works on all OS)
That way we could easily control the scaling of the content on the screen to prevent it from looking small and making it more comfortable for users.
And maybe an extra option to define if the scaling will be done by the operating system or AIR (since the "zoom" scaling inherited from Flash Player tends to be very resource intensive).
Also, it would be cool to add requestedDisplayResolution support for android.

**Comment by itlancer** [comment]
@LilithRainbows 
May be something similar already implemented via https://airsdk.dev/reference/actionscript/3.0/flash/display/Stage.html#fullScreenSourceRect but not for all platforms https://github.com/airsdk/Adobe-Runtime-Support/discussions/2422
Also related issue: https://github.com/airsdk/Adobe-Runtime-Support/issues/528

**Comment by LilithRainbows** [comment]
@itlancer 
Exactly, the idea would be a global option, that works even when the app is not in full screen.
Unfortunately the fullScreenSourceRect function does not work for Android either 😭

**Comment by ajwfrost** [comment]
A lot of how that works is just legacy left over from the Flash Player .. so when you say "whole Stage content", what do you actually mean? When you create a SWF, there is a stage defined at a certain size, and that size gets stored in the SWF structure and is used when generating the initial display. Which doesn't make a huge amount of sense on mobile phones! Hence the scale modes and resize events and the possibility to have the content react to the device display size and then to manually lay out its content.

Depending on your content and resolutions, if we just did scaling from the OS, then it could end up with very pixelated outputs e.g. if your stage was a lot smaller than the device resolution, the OS has to upscale it, and even with smoothing filters applied, this isn't ideal.

So I'm not quite sure how we would be able to improve upon that aspect of it -> happy to discuss further to see if there's a general approach we could use that would help though? (And yes, fullScreenSourceRect is an option, albeit with some similar issues and compatibility). It may be we need to draw some diagrams so that we can see what you expect to happen in various scenarios.

The requested display resolution for Android is something we're looking to implement via the ScreenMode class.

**Comment by LilithRainbows** [comment]
@ajwfrost 
Thanks for answering.
I think the static declaration of Stage size in the SWF header doesn't make much sense (at least in AIR).
That is why the StageScaleMode is mostly defined as noScale and the StageAlign as TL, to obtain a more responsive and adaptable design, which uses all the available space regardless of the screen sizes, nor the size declared in the SWF header.
The problem mainly occurs with the difference in screen densities, as unscaled content can look extremely small, that could normally be handled using the requestedDisplayResolution descriptor but it is not supported on all systems, as well as the fullScreenSourceRect function but it is not supported on all systems either and is limited to full screen.
By adding a custom scaling function (independent of the static size declared in the SWF) we could be sure of being able to control how much the content on the screen should be scaled (either for hdpi reasons or for user visual preference).
I also think that if possible we should have the option to choose whether the scaling will be handled by AIR or by the operating system, I know that the result can be pixelated, but the difference in performance can be very noticeable, and depending on the style of the game/app and os it can even look more visually beautiful, after all it would only be one more option available to be used in case the developer considers it more convenient.

**Comment by LilithRainbows** [comment]
Some temporary (and bad) options to scale content in android without resorting to external dependencies are:
1) Define in the SWF header a resolution lower than the device (respecting the aspect ratio) and define the StageScaleMode as exactFit.
Cons:
As it is a predefined static value, it would only work for a specific device model.
Performance drops considerably because scaling is handled by AIR.
2) Make use of the "android:largestWidthLimitDp" attribute to enable display compatibility mode.
Cons:
We have no control over how big the content will look and it could even make the situation worse depending on the device.

**Comment by MalacTheLittle** [comment]
I'm not sure if the scale of the stage content would be helpful since screens vary in aspect ratios and physical size. I solve this problem with `Capabilities.screenDPI` (more or less), focusing on scaling each individual object rather than the whole stage. The only problem I'm facing is that Android doesn't fetch the precise DPI like iOS, so the result is not as accurate as on iOS devices, but in most cases it meets the needs.

**Comment by LilithRainbows** [comment]
If we had the possibility to define a custom value for the stage size/scaling, we could take into account the different variables that we want to determine that value (such as screen density, resolution and user preference)
That way saves us from having to do it individually for each object, which would be a nightmare for very large projects that were already developed.

**Comment by 2jfw** [comment]
> If we had the possibility to define a custom value for the stage size/scaling, we could take into account the different variables that we want to determine that value (such as screen density, resolution and user preference) That way saves us from having to do it individually for each object, which would be a nightmare for very large projects that were already developed.

Did you consider the approach to put all your display list elements in a single sprite on stage/root rather than adding them to the stage directly and then scale that sprite?
